File: google_doc/lib/feature/auth/auth_controller.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:google_doc/services/storage_service.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'auth_repository.dart';
import '../../models/user_model.dart';

final authControllerProvider =
    AsyncNotifierProvider<AuthController, UserModel?>(AuthController.new);

class AuthController extends AsyncNotifier<UserModel?> {
  late final AuthRepository _repo;
  final GoogleSignIn _googleSignIn = GoogleSignIn(
    clientId: kIsWeb
        ? '272866144375-avm3n0kpui1ldft0uluef0bfj7h0a2ro.apps.googleusercontent.com'
        : null,
    scopes: ['email', 'profile'],
  );
  var _storage;
  bool _isAuthenticating = false;

  @override
  Future<UserModel?> build() async {
    _repo = ref.read(authRepositoryProvider);
    _storage = ref.read(storageServiceProvider);
  }

  // 🔵 GOOGLE SIGN IN
  Future<void> signInWithGoogle() async {
    if (_isAuthenticating) return;
    _isAuthenticating = true;

    state = const AsyncLoading();

    try {
      // In 6.2.1 on Web, you can use either .signIn() or .authenticate()
      // .authenticate() is better for modern FedCM/GIS compliance
      final GoogleSignInAccount? googleUser = await _googleSignIn.signIn();

      if (googleUser != null) {
        await _handleGoogleUser(googleUser);
      } else {
        // state = AsyncData(await _loadUserFromStorage());
      }
    } catch (e) {
      state = AsyncValue.error(e, StackTrace.current);
    } finally {
      _isAuthenticating = false;
    }
  }

  Future<void> _handleGoogleUser(GoogleSignInAccount account) async {
    try {
      final auth = await account.authentication;
      final idToken = auth.idToken;

      if (idToken == null) throw Exception('Google ID token missing');

      final user = await _repo.googleLogin(idToken);
      if (user.token != null) await _storage.saveToken(user.token!);

      state = AsyncData(user);
    } catch (e) {
      state = AsyncValue.error(e, StackTrace.current);
    }
  }

  Future<void> logout() async {
    await _googleSignIn.signOut();
    await _storage.deleteToken();
    state = const AsyncData(null);
  }

  // 🔐 LOGIN
  Future<void> login(String email, String password) async {
    state = const AsyncLoading();
    state = await AsyncValue.guard(() {
      return _repo.login(email, password);
    });
  }

  // 📝 REGISTER
  Future<void> register(String name, String email, String password) async {
    state = const AsyncLoading();
    state = await AsyncValue.guard(() {
      return _repo.register(name, email, password);
    });
  }
}

File: docs-backend/src/controller/auth.controller.js
import * as authService from "../service/auth.service.js";

export const register = async (req, res) => {
  const { name, email, password } = req.body;

  if (!name || !email || !password) {
    return res.status(400).json({ message: "Missing fields" });
  }

  try {
    const { token, user } = await authService.registerUser({ name, email, password });
    res.status(201).json({
      token,
      user: { id: user._id, name: user.name, email: user.email },
    });
  } catch (error) {
    if (error.message === "User already exists") {
      return res.status(400).json({ message: error.message });
    }
    console.error("Register Error:", error);
    res.status(500).json({ message: "Server error" });
  }
};

export const login = async (req, res) => {
  const { email, password } = req.body;

  if (!email || !password) {
    return res.status(400).json({ message: "Missing fields" });
  }

  try {
    const { token, user } = await authService.loginUser({ email, password });
    res.status(200).json({
      token,
      user: { id: user._id, name: user.name, email: user.email },
    });
  } catch (error) {
    if (error.message === "Invalid credentials") {
      return res.status(401).json({ message: error.message });
    }
    console.error("Login Error:", error);
    res.status(500).json({ message: "Server error" });
  }
};

export const googleAuth = async (req, res) => {
  const { idToken } = req.body;
  if (!idToken) {
    return res.status(400).json({ message: "Missing Google Token" });
  }
  try {
    const { token, user } = await authService.googleLogin(idToken);
    res.status(200).json({ token, user });
  } catch (error) {
    console.error("Google Auth Error:", error);
    res.status(500).json({ message: "Google authentication failed" });
  }
};

export const getUserProfile = async (req, res) => {
  try {
    const user = await authService.getUserById(req.user.id);
    res.status(200).json(user);
  } catch (error) {
    if (error.message === "User not found") {
      return res.status(404).json({ message: error.message });
    }
    console.error("Profile Error:", error);
    res.status(500).json({ message: "Server error" });
  }
};